distributed/worker_state_machine.py

```diff
@@ -3147,15 +3117,27 @@ def _validate_task_missing(self, ts: TaskState) -> None:
         assert ts in self.missing_dep_flight
 
     def _validate_task_cancelled(self, ts: TaskState) -> None:
-        assert ts.key not in self.data
-        assert ts.previous in {"long-running", "executing", "flight"}
-        # We'll always transition to released after it is done
         assert ts.next is None
+        assert ts.key not in self.data
+        if ts.previous in ("executing", "long-running"):
+            # Task dependencies may have been released;
+            # can't call _validate_task_executing
+            assert not ts.waiting_for_data
```

Comment:
@fjetter 

If you call `_validate_task_executing()` from here, only one test in the whole suite becomes flaky, `test_cancel_fire_and_forget` (dependencies can be released if the dependent is cancelled, but they must be in memory if it's executing) and on my computer only 0.1% of the times (more frequently on CI).
I added `test_cancel_with_dependencies_in_memory` to reproduce the use case deterministically. Still I find it very scary how little coverage we have for cancelled/resumed tasks with dependencies.

Reply:
> I find it very scary how little coverage we have for cancelled/resumed tasks with dependencies.

I do not disagree. At the time we started patching up the worker we barely had any coverage at all for any of these internals. We've come a long way and there is still a lot to do to reach 100% coverage.
We have to be a bit pragmatic here; I do not believe it is feasible or effective for us right now to push for absolute coverage.

Reply:
Follow-up: #6893